eth-contracts/contracts/service/ServiceProviderFactory.sol

```diff
@@ -348,9 +349,62 @@ contract ServiceProviderFactory is RegistryContract {
         return spId;
     }
 
-  /**
-   * @notice Update service provider balance
-   */
+    function addServiceType(
+        bytes32 _serviceType,
+        uint _serviceTypeMin,
+        uint _serviceTypeMax
+    ) external
+    {
+      require(
+          msg.sender == deployerAddress || msg.sender == registry.getContract(governanceKey),
+          "Only deployer or governance");
+      require(!this.isValidServiceType(_serviceType), "Already known service type");
+      validServiceTypes.push(_serviceType);
+      serviceTypeStakeRequirements[_serviceType] = ServiceInstanceStakeRequirements({
+          minStake: _serviceTypeMin,
+          maxStake: _serviceTypeMax
+      });
+    }
+
+    function removeServiceType(bytes32 _serviceType) external {
+      require(
+          msg.sender == deployerAddress || msg.sender == registry.getContract(governanceKey),
+          "Only deployer or governance");
+      uint serviceIndex = 0;
+      bool foundService = false;
+      for (uint i = 0; i < validServiceTypes.length; i ++) {
```

Comment:
you could avoid this iteration if you changed the `serviceTypeStakeRequirements` mapping to hold the index of the serviceType into the `validServiceTypes` array. so if you made it `serviceTypeInfo` mapping (bytes32 => struct ServiceTypeInfo` which held `{ index, minStake, maxStake }` then you could replace this iteration with just a check for existence in the mapping.

Reply:
yes, this is a good point - let's do it now

Reply:
also i want to apply this pattern to delegate ops if possible